pkg/buildpack/buildpack.go

```diff
@@ -326,6 +355,173 @@ func ToLayerTar(dest string, module BuildModule) (string, error) {
 	return layerTar, nil
 }
 
+func ToNLayerTar(dest string, module BuildModule) ([]ModuleTar, error) {
+	if !module.ContainsFlattenedModules() {
+		return handleSingleOrEmptyModule(dest, module)
+	}
+
+	modReader, err := module.Open()
+	if err != nil {
+		return nil, errors.Wrap(err, "opening blob")
+	}
+	defer modReader.Close()
+
+	tarCollection := newModuleTarCollection(dest)
+	tr := tar.NewReader(modReader)
+
+	// read the first header
+	header, err := tr.Next()
+	if err != nil {
+		if err == io.EOF {
+			return handleSingleOrEmptyModule(dest, module)
+		}
+		return nil, fmt.Errorf("failed to read first header '%s': %w", header.Name, err)
+	}
+
+	// the original version should be blank because the first header should look like "/cnb/buildpacks/<buildpack-id>"
+	origID, origVersion := parseBpIDAndVersion(header)
+	if origVersion != "" {
+		return nil, fmt.Errorf("first header '%s' contained unexpected version", header.Name)
+	}
+
+	if err := toNLayerTar(origID, origVersion, header, tr, tarCollection); err != nil {
+		return nil, err
+	}
+
+	errs := tarCollection.close()
+	if len(errs) > 0 {
+		return nil, errors.New("closing files")
+	}
+
+	return tarCollection.moduleTars(), nil
+}
+
+func toNLayerTar(origID, origVersion string, firstHeader *tar.Header, tr *tar.Reader, tc *moduleTarCollection) error {
+	toWrite := []*tar.Header{firstHeader}
+	if runtime.GOOS == "windows" && !strings.Contains(firstHeader.Name, "/cnb/buildpacks/") {
+		// On windows we can see the following headers before finding */cnb/buildpacks/<buildpack-id>/*
+		// Files
+		// Hives
+		// Files/cnb
+		// Files/cnb/buildpacks
+		for {
+			next, err := tr.Next()
+			if err != nil {
+				return err
+			}
+			if !strings.Contains(next.Name, "/cnb/buildpacks/") {
+				toWrite = append(toWrite, next)
+			} else {
+				origID, origVersion = parseBpIDAndVersion(next)
+				tc.buffer = toWrite
```

Comment:
Why do we need `buffer` to live on the tar collection instead of being a local variable? It is only used in this function...